processing/src/main/java/org/apache/druid/math/expr/ExprEval.java

```diff
@@ -644,6 +644,27 @@ public static Number computeNumber(@Nullable String value)
     return rv;
   }
 
+  /**
+   * Returns true if an {@link ExprEval} which has been cast to some other type is not equal to the original type.
+   * Effectively, this only happens with casts to {@link ExpressionType#LONG} and {@link ExpressionType#LONG_ARRAY}.
```

Comment:
That's a good point, I guess I was focused on casting to longs, and was having some trouble coming up with a name.

I've reworked things a bit and now the method is actually doing the casting, `ExprEval.castForComparison`, and is used more widely for building value matchers/predicates so if we decide to check for other things that should result in short-circuit of equality comparison we won't have to do much other than update this method. I think this probably fixed a few problems with equality filters with string match value types that are not convertible to numbers, but I haven't added tests yet. I have not wired this method up for range filters though, because at least for the double to long case, we can still do comparison, just not directly with the cast values. I need to think a bit more about this. Maybe i should put 'equality' in the method name since that is the only comparison it is currently used for?

Reply:
I do think it makes sense to put "equality" in the name. Rationale: for the examples you mentioned, like comparing double `1.1` using match type long, the logic needs to change if it's a range filter vs an equality filter. (`1.1` can never be equal to a long, but it can be greater than or less than one)

Reply:
cool, it is now `castForEqualityComparison`